Submit AYON_SERVER_URL to jobs

BigRoy opened the pull request:
## Changelog Description

Submit `AYON_SERVER_URL` with the Deadline Job.

To avoid submitting along the user's API key or exposing the service user API key in AYON settings we only submit along the `AYON_SERVER_URL` however we now have an extra setting on the Deadline repository for the AYON Server URL submitted with the jobs to supply their API keys.

## Additional info

![image](https://github.com/user-attachments/assets/e6a46ffa-78ef-4eba-97bf-8c787917fa22)

![image](https://github.com/user-attachments/assets/0a3e40ce-e0bf-43b2-b459-d2c4c8866be5)

Fixes https://github.com/ynput/ayon-deadline/issues/27

⚠️ This requires updating the GlobalJobPreLoad and the AYON plug-in on Deadline Repository

Documentation with use case description https://github.com/ynput/ayon-documentation/pull/285

## Testing notes:

1. Enable the plug-in in settings: `ayon+settings://deadline/publish/CollectAYONServerToFarmJob/enabled`

![image](https://github.com/user-attachments/assets/0a3e40ce-e0bf-43b2-b459-d2c4c8866be5)

2. Configure additional server URL if you have a custom server URL (other than defined in defaults in AYON settings)
3. Submitting deadline jobs will submit along `AYON_SERVER_URL` if plug-in is enabled.
4. The GlobalJobPreLoad plug-in will resolve the API key for a custom server URL submitted along as defined in its plugin settings.

BigRoy commented on client/ayon_deadline/plugins/publish/global/collect_deadline_job_env_vars.py:
```diff
@@ -0,0 +1,78 @@
+        "PYTHONPATH",
+        "PATH",
+    ]
+
+
+class CollectAYONServerUrlToFarmJob(CollectDeadlineJobEnvVars):
+    label = "Submit AYON server URL to farm job"
+
+    # TODO: Expose the enabled/optional state to settings
+    ENV_KEYS = [
+        "AYON_SERVER_URL",
+        "AYON_USE_DEV",
+    ]
```
### Review only this code if doing code review

**NOTE** 

This is the ONLY related code changes for this PR - everything else is also in #32

kalisp replied:
Do we want send AYON_SERVER_URL always? If thats the case, server url in AYON Deadline plugin doesn't make sense, neither code for it in `GlobalJobPreload`.

(Just note, API keys for all DL servers would need to be same as they would be still set via AYON DL plugin.)

BigRoy replied:
> Do we want send AYON_SERVER_URL always? If thats the case, server url in AYON Deadline plugin doesn't make sense, neither code for it in GlobalJobPreload.

No - I think we'd want to make this plug-in disabled by default, and configurable via settings to enable it.

> (Just note, API keys for all DL servers would need to be same as they would be still set via AYON DL plugin.)

Sir - that's a **very valid** point. We may need to submit that along too. @tweak-wtf, thoughts?

tweak-wtf replied:
oh sheet, true! thanks for catching!
i really like what's described in https://github.com/ynput/ayon-documentation/pull/285
seems like minimal impact on the existing addon codebase and on the additional AYON instances settings.

if we want to expand on that and the n:n idea between AYON and DL, i could imagine the AYON Deadline Plugin Settings getting either a new section `Additional AYON Servers` or a general list of dicts kinda like the current AYON server settings for the DL servers.
then submitting the current AYON URL could become enabled by default. whether that's desirable, i can't really say.

for me using `AYON_API_KEY` in `ayon+settings://core/environments` as an override in that instance's settings really does the trick.

BigRoy replied:
How does that work, with the `AYON_API_KEY` in settings? When `GlobalJobPreLoad` triggers the connection with the AYON server - it already needs a valid API key to make the connection? Seems like for you it may just happen to work because it maybe found some matching 'login' using your local cached login?

BigRoy commented:
Tagging @tweak-wtf who might be interested in this.

kalisp commented:
Env var is being passed, but Settings for that collector should be added to make it optional.

kalisp commented:
Is `Submit AYON server URL to farm job` proper label as it seems it could push AYON_API_KEY also. I am guessing it we need to push different AYON_API_KEY value, it should be set in global environment variables in `ayon+settings://core/environments`?

![image](https://github.com/user-attachments/assets/24a869d9-d528-45e3-8270-3092a4f0bad1)
(I was thinking about usage during writing small documentation https://github.com/ynput/ayon-documentation/pull/285 )

BigRoy commented:
> Is `Submit AYON server URL to farm job` proper label as it seems it could push AYON_API_KEY also. I am guessing it we need to push different AYON_API_KEY value, it should be set in global environment variables in `ayon+settings://core/environments`?

Agree. What if we generalize this further, and make the settings like a `dict` where one can specify any amount of keys? The default settings could have `AYON_API_KEY` and `AYON_SERVER_URL` defined, but the plugin just disabled. That way it is as easy as just enabling it by default?

However, it wouldn't really clarify why you'd want to pass that.

Ideas @antirotor @m-u-r-p-h-y @iLLiCiTiT - better naming? different approach?

iLLiCiTiT commented on server/settings/publish_plugins.py:
```diff
@@ -380,6 +390,11 @@ class PublishPluginsModel(BaseSettingsModel):
         "primary_pool": "",
         "secondary_pool": ""
     },
+    "CollectAYONServerUrlToFarmJob": {
```
Wouldn't be better to have settings for `CollectDeadlineJobEnvVars` with enum `Pass AYON server url to job` and options `Always, User choice, Never` instead of new plugin?

iLLiCiTiT commented on client/ayon_deadline/plugins/publish/global/collect_deadline_job_env_vars.py:
```diff
@@ -53,3 +51,17 @@ def process(self, context):
+
+    ENV_KEYS = [
+        "AYON_SERVER_URL",
+        "AYON_API_KEY",
```
I don't like to pass api key to the job, this is huge blocker.

BigRoy replied:
Valid point - the implementation here is mostly for DEV servers. I would like it, and @tweak-wtf too. We could of course mark it as such and put a description on settings that it's intended exactly for that?

And.. a better improvement would be to define in AYON settings for this plug-in which API Key to send along? So that you can still set it to the service user API key. Would that solve your concerns?

kalisp replied:
`AYON_API_KEY` is not artist api key, it should be service API key for connecting DL to AYON.  
It could be there, if one would need providing API key value different for each Deadline. In that case value currently set in Deadline AYON plugin wouldnt be enough.
![image](https://github.com/user-attachments/assets/5768c66f-f5d1-441c-8430-fdcf6e030ed3)

BigRoy replied:
> `AYON_API_KEY` is not artist api key, it should be service API key for connecting DL to AYON. It could be there, if one would need providing API key value different for each Deadline. In that case value currently set in Deadline AYON plugin wouldnt be enough.

I suppose you're replying to @iLLiCiTiT 's comment. So you agree with having `AYON_API_KEY` to be specified in AYON studio/project settings for this particular addon would be a good workaround @kalisp ?

kalisp replied:
I was testing it by adding `AYON_API_KEY` to `ayon+settings://core/project_environments?project=test_project_1` and it was working. 

I think this is enough for such a edge case. 

I just dont like that `Submit ...` label.

I feel neither here nor there about Kuba comment to smush it into single collector `CollectDeadlineJobEnvVars`

![image](https://github.com/user-attachments/assets/b02842d0-a30e-49c1-b39b-a193ab2aedb6)

![image](https://github.com/user-attachments/assets/99415fc3-b820-41d6-b889-7dc7621d5d9b)

BigRoy replied:
> I was testing it by adding `AYON_API_KEY` to `ayon+settings://core/project_environments?project=test_project_1` and it was working.
> (Without it it was picking value from DL AYON plugin correctly, was worried empty value would override it.)

This couldn't actually work on the `GlobalJobPreLoad` because it needs the API key to connect to the server first - so it can't access the settings to get that API key in the first place. :)

kalisp replied:
What? In `GlobalJobPreLoad` it now checks if job has that env var, if it does, it uses it, if not, it looks for api key from DL plugin.

And that is how it is actually supposed to work. (Or I misunderstood.)
![image](https://github.com/user-attachments/assets/a5339f5f-89f7-4eeb-ac75-40e144af1dfa)

iLLiCiTiT replied:
This adds AYON_API_KEY to the job during publishing. Which means it adds artist api key.

iLLiCiTiT replied:
> I was testing it by adding AYON_API_KEY to ayon+settings://core/project_environments?project=test_project_1 and it was working.

How do you access project settings without api key? You must have artist api key passed to job to be able to do it.

kalisp replied:
Yes, I am stupid, I was sure that when I tested it for the first time, logging of AYON_API_KEY was empty in that collector, so I thought that env var is unoccupied in regular AYON, until its is provided by global environment.

It seems its not the case, it holds actually ARTIST key by default.

But new collector is not replacing it, just reusing it, if there is no value in Settings.

Note about providing API key is in hover message and it will be in documentation, so I think it should be fine.

iLLiCiTiT commented on client/ayon_deadline/plugins/publish/global/collect_deadline_job_env_vars.py:
```diff
@@ -53,3 +51,17 @@ def process(self, context):
+    ENV_KEYS = [
+        "AYON_SERVER_URL",
+        "AYON_API_KEY",
+        "AYON_USE_DEV",
```
Why do we need to pass this? AYON_BUNDLE_NAME covers that, or?

BigRoy replied:
That was my question - `AYON_USE_DEV` is hence redundant? We need the SERVER URL and API KEY to allow a single Deadline server to support submissions from multiple AYON servers (mostly for 'dev' usage currently).

iLLiCiTiT replied:
So this whole PR is adding new plugin and settings for dev?

kalisp reviewed (approved):
Tested, worked.

My objection is to `Submit AYON server URL to farm job`, but otherwise fine.

BigRoy commented:
@kalisp @iLLiCiTiT can you check the code and functionality again?

![image](https://github.com/user-attachments/assets/ff26aff6-936c-4d6a-9e5b-0b717be28f4c)
![image](https://github.com/user-attachments/assets/b2bc723a-ee13-4eaa-bb9d-1ae1d83913b4)
![image](https://github.com/user-attachments/assets/c20f29c1-29c5-4d1a-8d68-7af39614a05e)

With API key set:
![image](https://github.com/user-attachments/assets/0ad41544-1a18-49d6-a370-6607a409c9c5)
![image](https://github.com/user-attachments/assets/867475f5-05ee-4102-830f-535c89e3a8c0)

iLLiCiTiT commented:
Can someone explain me what is this PR about? Passing server ulr is weird, passing api key is big No, No. What is purpose of this PR? From the conversations it sounds like we're adding plugin and settings for developers... Flooding studios with plugins and settings they should not touch is not the way.

BigRoy commented:
> Shouldn't we allow to define more api keys per server in the AYON plugin then? So we pass in only server url and plugin will decide? Also keep in mind that AYON url on deadline does not have to be the same as what artists use.

So what would GlobalJobPreLoad do... just start connecting to each server one by one? Since we're not submitting the AYON SERVER URL by default either.

BigRoy commented:
> So we pass in only server url and plugin will decide? Also keep in mind that AYON url on deadline does not have to be the same as what artists use.

How does that work? Why shouldn't it - how would what I submit still correspond to tasks/assets on that server? Or do you mean that they might be on a different network that accesses the same server through a different URL?

iLLiCiTiT commented:
If `AYON_SERVER_URL` is set then find it in the list. If is not set, just use the first one. If is not found, then it is question if we should raise error or try each api key from plugin config to connect to the server.

BigRoy commented:
> If `AYON_SERVER_URL` is set then find it in the list. If is not set, just use the first one. If is not found, then it is question if we should raise error or try each api key from plugin config to connect to the server.

I see. That still wouldn't solve this either:

> Also keep in mind that AYON url on deadline does not have to be the same as what artists use.

As such... I wonder if that design is good either. However, I think it's a valid point and we may need to think of a different strategy.
I'll mark this as draft again just to clarify it's up for discussion still.

iLLiCiTiT commented:
If we need this ASAP for a client I'm kinda ok with having this. BTW I would not add new plugin and would keep it in global one as `Pass server url`. That would now also pass in AYON API if enabled.

iLLiCiTiT commented on client/ayon_deadline/plugins/publish/global/collect_deadline_job_env_vars.py:
```diff
@@ -52,4 +50,4 @@ def process(self, context):
 
         # Transfer some environment variables from current context
-        context.data.setdefault(JOB_ENV_DATA_KEY, {}).update(env)
+        context.data.setdefault(JOB_ENV_DATA_KEY, {}).update(env)
```
```suggestion
        context.data.setdefault(JOB_ENV_DATA_KEY, {}).update(env)

```

iLLiCiTiT commented on server/settings/publish_plugins.py:
```diff
@@ -7,6 +7,23 @@
 
 
+class CollectAYONServerToFarmJobModel(BaseSettingsModel):
+    enabled: bool = SettingsField(True, title="Enabled")
```
```suggestion
    enabled: bool = SettingsField(False, title="Enabled")
```

iLLiCiTiT commented on server/settings/publish_plugins.py:
```diff
@@ -7,6 +7,23 @@
 
+class CollectAYONServerToFarmJobModel(BaseSettingsModel):
+    enabled: bool = SettingsField(True, title="Enabled")
+    optional: bool = SettingsField(True, title="Optional")
+    active: bool = SettingsField(True, title="Active")
```
```suggestion|
```

iLLiCiTiT commented on server/settings/publish_plugins.py:
```diff
@@ -380,6 +408,12 @@ class PublishPluginsModel(BaseSettingsModel):
+    "CollectAYONServerUrlToFarmJob": {
+        "enabled": False,
+        "active": True,
+        "optional": False,
```
```suggestion
```

iLLiCiTiT commented on client/ayon_deadline/plugins/publish/global/collect_deadline_job_env_vars.py:
```diff
@@ -44,6 +39,9 @@ class CollectDeadlineJobEnvVars(pyblish.api.ContextPlugin):
     ]
 
     def process(self, context):
+        if not self.is_active(context.data):
+            return
+
```
```suggestion
```

MustafaJafar commented on client/ayon_deadline/plugins/publish/global/collect_deadline_job_env_vars.py:
```diff
@@ -1,18 +1,13 @@
 
-class CollectDeadlineJobEnvVars(pyblish.api.ContextPlugin):
+class CollectDeadlineJobEnvVars(pyblish.api.ContextPlugin,
+                                OptionalPyblishPluginMixin):
```
why is it `OptionalPyblishPluginMixin` although it doesn't have settings?

MustafaJafar reviewed (commented):
Sorry, I didn't go through all of the comments. 
Anyways, This PR works and `AYON_API_KEY` is not propagated to the publish job when the collector is enabled.

BigRoy commented:
@kalisp @iLLiCiTiT I've implemented the implementation that @iLLiCiTiT proposed in the call we had about this PR. Now the `AYON_API_KEY` is to be configured is _not_ configured in the AYON Settings on the server nor is it submitted along with the job. Instead only the server URL is submitted along but the API key is then resolved on the GlobalJobPreLoad by having server URLs configured on the deadline repository AYON plug-in settings. The PR description is also updated.

I personally like it less. It feels overly complicated. And only after I implemented this I figured this would also have problems if say two developers on one network are connecting to their servers using e.g. `localhost` as server URL, submit to the farm (to run only on their machine, because it's their dev job for example) then it still couldn't resolve the relevant API key unless both users use the same service user API key. Again, an edge case. My biggest concern is the 'somewhat more' complexity involved.

Anyway, please review and comment so we can decide to continue this way, or revert to the previous logic of this PR or do it even differently.

@tweak-wtf feel free to let you know your thoughts too.

tweak-wtf commented:
if i understood correctly we'd now define additional AYON servers and their corresponding API keys in the AYON Deadline Plugin with this syntax `<AYON_URL>:<AYON_PORT>@<AYON_API_KEY>` alongside the default `AYON Credentials` settings resembling the main AYON server.
so we're giving the API key as a string to the Deadline Plugin as opposed to passing them around in dictionaries from a publish process. wouldn't that also expose the API KEy which we're trying to prevent?

However, this sounds fine to me really. I understand the security concerns that passing API keys in cleartext is badbadnotgood.

just a question... would it be possible/feasible to e.g. reference a secret name present in the main ayon instance instead for the actual API key? so something like `dev.awesome.ayon.app@ayon-instance-dev-key-1`?
imho this could make it even more secure since the keys are stored as secrets on the main instance and would only be retrieved from `GlobalJobPreLoad` if a job was submitted from one of the defined extra servers. 

btw. from my side this is not urgent at all. i just think it'd be a great addition for devs to further isolate a prod env from a sandbox and could ease some pain in transitional phases from one ayon instance to another

BigRoy commented:
> so we're giving the API key as a string to the Deadline Plugin as opposed to passing them around in dictionaries from a publish process. wouldn't that also expose the API KEy which we're trying to prevent?

It does - but only to what's supposedly the admin/super-user on Deadline.

> would it be possible/feasible to e.g. reference a secret name present in the main ayon instance instead for the actual API key?

I'd say no - because you'd first need to connect to the instance ;) using... an API key.

tweak-wtf commented:
got ya. then it really feels fine to me.

iLLiCiTiT commented on client/ayon_deadline/repository/custom/plugins/GlobalJobPreLoad.py:
```diff
@@ -13,7 +13,7 @@
     DirectoryUtils,
 )
-__version__ = "1.1.3"
+__version__ = "1.1.4"
```
This deserves minor bump.
```suggestion
__version__ = "1.2.0"
```

BigRoy commented:
@iLLiCiTiT should this also get bump minor label?

iLLiCiTiT commented:
> @iLLiCiTiT should this also get bump minor label?

Yes please.

MustafaJafar reviewed (approved):
This is cool, it works on my side. 
I love the new idea of having Multiline AYON servers field. 
I was too lazy to run an additional AYON instance so I've just removed the AYON server URL and used the new field solely. which makes me question the existence of the old setting. But anyways, this works!.
![image](https://github.com/user-attachments/assets/b3235a5e-7c42-4011-9433-f353495c6263)

kalisp commented:
I added additional casting to list as I was suffering with:
![image](https://github.com/user-attachments/assets/52626f9c-d7dc-497c-a4b6-0d48af8f5a94)
